In [2]:
######################################################
#    FIS616 Introdução à Computação em Física
#
#                   01/2020
#
#           
#           
#           Prof. V.B. Nascimento
#                   @UFMG
#
#  
######################################################

######################################################################################################
# ALUNO : JOAO LUCAS RODRIGUES
# MATRÍCULA : 2018116112
# TURMA : TM
######################################################################################################

######################################################
#
#               Atividade Avaliativa 10 
#      
#              Algoritmos de ordenamento
######################################################

In [2]:
# Bubble Sort
def bubblesort(list):

# Swap the elements to arrange in order
    for iter_num in range(len(list)-1,0,-1):
        for idx in range(iter_num):
            if list[idx]>list[idx+1]:
                temp = list[idx]
                list[idx] = list[idx+1]
                list[idx+1] = temp


list = [19,2,31,45,6,11,121,27]
bubblesort(list)
print(list)

[2, 6, 11, 19, 27, 31, 45, 121]


In [3]:
# Merge Sort

def merge_sort(unsorted_list):
    if len(unsorted_list) <= 1:
        return unsorted_list
# Find the middle point and devide it
    middle = len(unsorted_list) // 2
    left_list = unsorted_list[:middle]
    right_list = unsorted_list[middle:]

    left_list = merge_sort(left_list)
    right_list = merge_sort(right_list)
    lista = merge(left_list, right_list)
    return lista 


# Merge the sorted halves

def merge(left_half,right_half):

    res = []
    while len(left_half) != 0 and len(right_half) != 0:
        if left_half[0] < right_half[0]:
            res.append(left_half[0])
            left_half.remove(left_half[0])
        else:
            res.append(right_half[0])
            right_half.remove(right_half[0])
    if len(left_half) == 0:
        res = res + right_half
    else:
        res = res + left_half
    return res

unsorted_list = [64, 34, 25, 12, 22, 11, 90]

print(merge_sort(unsorted_list))

[11, 12, 22, 25, 34, 64, 90]


In [4]:
# Insertion Sort
def insertion_sort(InputList):
    for i in range(1, len(InputList)):
        j = i-1
        nxt_element = InputList[i]
# Compare the current element with next one

        while (InputList[j] > nxt_element) and (j >= 0):
            InputList[j+1] = InputList[j]
            j=j-1
        InputList[j+1] = nxt_element

list = [19,2,31,45,30,11,121,27]
insertion_sort(list)
print(list)

[2, 11, 19, 27, 30, 31, 45, 121]


In [5]:
# Shell Sort
def shellSort(input_list):
    
    gap = len(input_list) // 2
    while gap > 0:

        for i in range(gap, len(input_list)):
            temp = input_list[i]
            j = i
# Sort the sub list for this gap

            while j >= gap and input_list[j - gap] > temp:
                input_list[j] = input_list[j - gap]
                j = j-gap
            input_list[j] = temp

# Reduce the gap for the next element

        gap = gap//2

list = [19,2,31,45,30,11,121,27]

shellSort(list)
print(list)

[2, 11, 19, 27, 30, 31, 45, 121]


In [6]:
# Selection Sort
def selection_sort(input_list):

    for idx in range(len(input_list)):

        min_idx = idx
        for j in range( idx +1, len(input_list)):
            if input_list[min_idx] > input_list[j]:
                min_idx = j
# Swap the minimum value with the compared value

        input_list[idx], input_list[min_idx] = input_list[min_idx], input_list[idx]


l = [19,2,31,45,30,11,121,27]
selection_sort(l)
print(l)

[2, 11, 19, 27, 30, 31, 45, 121]


In [7]:
import numpy
import random 
import time

In [15]:
inicio = time.time()
V = []
U = []
W = []
Z = []
for i in range(100):
    a = random.uniform(0,100)
    V.append(a)
for i in range(1000):
    b = random.uniform(0,1000)
    U.append(b)
for i in range(10000):
    c = random.uniform(0,10000)
    V.append(c)
for i in range(100000):
    d = random.randint(0,10000)
    Z.append(d)
    
fim = time.time()
print('\n tempo para gerar os conjuntos de vetores aleatorios:\n',fim-inicio)


 tempo para gerar os conjuntos de vetores aleatorios:
 0.20218992233276367


In [16]:
#avaliação 5 metodos 100
inicio = time.time()
bubblesort(V)
fim = time.time()
print('Bubble sort:', fim - inicio)

inicio = time.time()
merge_sort(V)
fim = time.time()
print('Merge sort:', fim - inicio)

inicio = time.time()
insertion_sort(V)
fim = time.time()
print('Insertion sort:', fim - inicio)

inicio = time.time()
shellSort(V)
fim = time.time()
print('Shell sort:', fim - inicio)

inicio = time.time()
selection_sort(V)
fim = time.time()
print('Selection sort:', fim - inicio)


Bubble sort: 9.367389917373657
Merge sort: 0.04080343246459961
Insertion sort: 0.0017743110656738281
Shell sort: 0.02201080322265625
Selection sort: 3.346344232559204


In [10]:
#avaliação 5 metodos 1000
inicio = time.time()
bubblesort(U)
fim = time.time()
print('bubblesort:', fim - inicio)

inicio = time.time()
merge_sort(U)
fim = time.time()
print('merge_sort:', fim - inicio)

inicio = time.time()
insertion_sort(U)
fim = time.time()
print('insertion_sort:', fim - inicio)

inicio = time.time()
shellSort(U)
fim = time.time()
print('shellSort:', fim - inicio)

inicio = time.time()
selection_sort(U)
fim = time.time()
print('selection_sort:', fim - inicio)

bubblesort: 0.0872189998626709
merge_sort: 0.0028934478759765625
insertion_sort: 0.00022029876708984375
shellSort: 0.0012176036834716797
selection_sort: 0.031119585037231445


In [17]:
#avaliação 5 metodos 10000
inicio = time.time()
bubblesort(W)
fim = time.time()
print('bubblesort:', fim - inicio)

inicio = time.time()
merge_sort(W)
fim = time.time()
print('merge_sort:', fim - inicio)

inicio = time.time()
insertion_sort(W)
fim = time.time()
print('insertion_sort:', fim - inicio)

inicio = time.time()
shellSort(W)
fim = time.time()
print('shellSort:', fim - inicio)

inicio = time.time()
selection_sort(W)
fim = time.time()
print('selection_sort:', fim - inicio)

bubblesort: 7.43865966796875e-05
merge_sort: 6.246566772460938e-05
insertion_sort: 6.103515625e-05
shellSort: 5.936622619628906e-05
selection_sort: 5.9604644775390625e-05


In [1]:
#avaliação 5 metodos 100000
inicio = time.time()
bubblesort(Z)
fim = time.time()
print('bubblesort:', fim - inicio)

inicio = time.time()
merge_sort(Z)
fim = time.time()
print('merge_sort:', fim - inicio)

inicio = time.time()
insertion_sort(Z)
fim = time.time()
print('insertion_sort:', fim - inicio)

inicio = time.time()
shellSort(Z)
fim = time.time()
print('shellSort:', fim - inicio)

inicio = time.time()
selection_sort(Z)
fim = time.time()
print('selection_sort:', fim - inicio)

#para averiguar o programa basta ir na opção "Kernel" e apertar "Restart and run all"

# Resultados

## 1º caso

### vetor de 100 elementos mais eficiente insertion_sort, e pior desempenho Bubble_Sort

Bubble sort: 9.367389917373657
Merge sort: 0.04080343246459961
Insertion sort: 0.0017743110656738281
Shell sort: 0.02201080322265625
Selection sort: 3.346344232559204

## 2º caso

### vetor de 1000 elementos mais eficiente insertion_sort, e pior desempenho Bubble_Sort

bubblesort: 0.0872189998626709
merge_sort: 0.0028934478759765625
insertion_sort: 0.00022029876708984375
shellSort: 0.0012176036834716797
selection_sort: 0.031119585037231445

## 3º caso

### vetor de 10000 elementos mais eficiente shellSort, e pior desempenho Bubble_Sort

bubblesort: 7.43865966796875e-05
merge_sort: 6.246566772460938e-05
insertion_sort: 6.103515625e-05
shellSort: 5.936622619628906e-05
selection_sort: 5.9604644775390625e-05

# Conclusão:

Nos casos acima é possível averiguar que  pior desempenho em todos os casos foi o Bubble_Sort 
Eu não o escolheria para um oordenamento grande, e quando o array cresce shell e selection se mostram mais eficientes que o até então insertion que se aproxima bastante dos resultados de merge